python/mxnet/utils.py

```diff
@@ -0,0 +1,151 @@
+
+    Parameters
+    ----------
+    filename : str
+        Path to the file.
+    sha1_hash : str
+        Expected sha1 hash in hexadecimal digits.
+
+    Returns
+    -------
+    bool
+        Whether the file content matches the expected hash.
+    """
+    sha1 = hashlib.sha1()
+    with open(filename, 'rb') as f:
+        while True:
+            data = f.read(1048576)
+            if not data:
+                break
+            sha1.update(data)
+
+    return sha1.hexdigest() == sha1_hash
+
+
+def _download_retry_with_backoff(fname, url, max_attempts=4):
+    """Downloads a url with backoff applied, automatically re-requesting after a failure.
+    :param fname : str
+        Name of the target file being downloaded.
+    :param url : str
+        URL to download.
+    :param max_attempts : int, optional
+        The number of requests to attempt before throwing a RequestException.
+    :return:
+        A streaming request object.
+    """
+    attempts = 1
+    backoff_coef = 50.0
+    while True:
+        try:
+            print('Downloading %s from %s...' % (fname, url))
+            r = requests.get(url, stream=True)
+
+            # If the remote server returned an error, raise a descriptive HTTPError.
+            # For non-error http codes (e.g. 200, 206) do nothing.
+            r.raise_for_status()
+            return r
+        except requests.exceptions.RequestException:
+            # Likely non-2** result, possibly timeout or redirection failure.
+            attempts = attempts + 1
+            if attempts > max_attempts:
+                print('Downloading %s from %s, failed after #%d attempts' % (fname, url, attempts))
+                raise
+
+            # Apply backoff with default values borrowed from the popular Boto3 lib.
+            time.sleep((backoff_coef * (2 ** attempts)) / 1000.0)
+
+
+def download(url, path=None, overwrite=False, sha1_hash=None):
+    """Download a given URL
+
+    Parameters
+    ----------
+    url : str
+        URL to download
+    path : str, optional
+        Destination path to store downloaded file. By default stores to the
+        current directory with same name as in url.
+    overwrite : bool, optional
+        Whether to overwrite destination file if already exists.
+    sha1_hash : str, optional
+        Expected sha1 hash in hexadecimal digits. Will ignore existing file when hash is specified
+        but doesn't match.
+
+    Returns
+    -------
+    str
+        The file path of the downloaded file.
+    """
+    if path is None:
+        fname = url.split('/')[-1]
+    elif os.path.isdir(path):
+        fname = os.path.join(path, url.split('/')[-1])
+    else:
+        fname = path
+
+    if overwrite or not os.path.exists(fname) or (sha1_hash and not check_sha1(fname, sha1_hash)):
+        dirname = os.path.dirname(os.path.abspath(os.path.expanduser(fname)))
+        if not os.path.exists(dirname):
```

Comment:
you can use exist_ok and avoid the check:

https://docs.python.org/3/library/os.html#os.makedirs

Reply:
that's py3 feature. https://docs.python.org/2.7/library/os.html#os.makedirs